django_fido/admin/authenticator.py

```diff
@@ -0,0 +1,108 @@
+"""Admin for django_fido authenticator."""
+from __future__ import unicode_literals
+
+from django import forms
+from django.conf.urls import url
+from django.contrib import admin
+from django.contrib.auth import get_user_model
+from django.contrib.auth.mixins import PermissionRequiredMixin
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.http import Http404, HttpResponse, HttpResponseRedirect
+from django.urls import reverse, reverse_lazy
+from django.utils.translation import ugettext_lazy as _
+
+from django_fido.forms import Fido2RegistrationForm
+from django_fido.models import Authenticator
+from django_fido.views import Fido2RegistrationRequestView, Fido2RegistrationView
+
+
+class Fido2RegistrationAdminForm(Fido2RegistrationForm):
+    """Registration form with user selection."""
+
+    user = forms.ModelChoiceField(
+        queryset=get_user_model().objects.all(),
+        error_messages={'required': _("Operation wasn't completed.")},
+    )
+
+    field_order = ('user', 'label')
+
+
+class Fido2RegistrationRequestAdminView(PermissionRequiredMixin, Fido2RegistrationRequestView):
+    """Registration request view with user selection."""
+
+    permission_required = 'django_fido.add_authenticator'
+
+    def get_user(self):
+        """Get user based on POST request."""
+        user_id = int(self.request.POST.get('user'))
+        try:
+            return get_user_model().objects.get(pk=user_id)
+        except get_user_model().DoesNotExist:
+            raise Http404('User does not exist')
+
+
+class AuthenticatorAddView(PermissionRequiredMixin, Fido2RegistrationView):
+    """Authenticator add view."""
+
+    permission_required = 'django_fido.add_authenticator'
+
+    form_class = Fido2RegistrationAdminForm
+    template_name = 'django_fido/add_authenticator.html'
+    fido2_request_url = reverse_lazy('admin:django_fido_registration_request')
+    extra_context = None
+
+    def form_valid(self, form: forms.Form) -> HttpResponse:
+        """Complete the FIDO token registration."""
+        try:
+            # Return value is ignored, because we need whole attestation.
+            self.complete_registration(form)
+        except ValidationError as error:
+            form.add_error(None, error)
+            return self.form_invalid(form)
+
+        authenticator = Authenticator.objects.create(
+            user=form.cleaned_data['user'],
+            attestation=form.cleaned_data['attestation'],
+            label=form.cleaned_data.get('label'),
+        )
+        return HttpResponseRedirect(reverse('admin:django_fido_authenticator_change', args=(authenticator.pk,)))
```

Comment:
Redirect should be handled by the `success_url`.

Reply:
The url is dependent on the `pk` of new object and this is not a `ModelFormView`, so it would be difficult to transfer the aforementioned `pk` to `get_success_url` method.

Reply:
OK, that would make sense for `ModelFormView`.